// @gooddata/typings/dist/src/tests/AFM.test.js
"use strict";
exports.__esModule = true;
// (C) 2007-2020 GoodData Corporation
var AFM_1 = require("../AFM");
describe('AFM', function () {
    describe('isObjectUriQualifier', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isObjectUriQualifier(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isObjectUriQualifier(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when identifier object qualifier is tested', function () {
            var objectQualifier = {
                identifier: 'id'
            };
            var result = AFM_1.AFM.isObjectUriQualifier(objectQualifier);
            expect(result).toEqual(false);
        });
        it('should return true when uri object qualifier is tested', function () {
            var objectQualifier = {
                uri: '/gdc/mock/id'
            };
            var result = AFM_1.AFM.isObjectUriQualifier(objectQualifier);
            expect(result).toEqual(true);
        });
    });
    describe('isObjIdentifierQualifier', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isObjIdentifierQualifier(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isObjIdentifierQualifier(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when uri object qualifier is tested', function () {
            var objectQualifier = {
                uri: '/gdc/mock/id'
            };
            var result = AFM_1.AFM.isObjIdentifierQualifier(objectQualifier);
            expect(result).toEqual(false);
        });
        it('should return true when identifier object qualifier is tested', function () {
            var objectQualifier = {
                identifier: 'id'
            };
            var result = AFM_1.AFM.isObjIdentifierQualifier(objectQualifier);
            expect(result).toEqual(true);
        });
    });
    describe('isSimpleMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isSimpleMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isSimpleMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when simple measure definition is tested', function () {
            var measure = {
                measure: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = AFM_1.AFM.isSimpleMeasureDefinition(measure);
            expect(result).toEqual(true);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measure = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = AFM_1.AFM.isSimpleMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return false when pop measure definition is tested', function () {
            var measure = {
                popMeasure: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = AFM_1.AFM.isSimpleMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measure = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = AFM_1.AFM.isSimpleMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
    });
    describe('isArithmeticMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isArithmeticMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isArithmeticMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measure = {
                measure: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = AFM_1.AFM.isArithmeticMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return true when arithmetic measure definition is tested', function () {
            var measure = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = AFM_1.AFM.isArithmeticMeasureDefinition(measure);
            expect(result).toEqual(true);
        });
        it('should return false when pop measure definition is tested', function () {
            var measure = {
                popMeasure: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = AFM_1.AFM.isArithmeticMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measure = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = AFM_1.AFM.isArithmeticMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
    });
    describe('isPopMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isPopMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isPopMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measure = {
                measure: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = AFM_1.AFM.isPopMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measure = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = AFM_1.AFM.isPopMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return true when pop measure definition is tested', function () {
            var measure = {
                popMeasure: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = AFM_1.AFM.isPopMeasureDefinition(measure);
            expect(result).toEqual(true);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measure = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = AFM_1.AFM.isSimpleMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
    });
    describe('isPreviousPeriodMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isPreviousPeriodMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isPreviousPeriodMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measure = {
                measure: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = AFM_1.AFM.isPreviousPeriodMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measure = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = AFM_1.AFM.isPreviousPeriodMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return false when pop measure definition is tested', function () {
            var measure = {
                popMeasure: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = AFM_1.AFM.isPreviousPeriodMeasureDefinition(measure);
            expect(result).toEqual(false);
        });
        it('should return true when previous period measure definition is tested', function () {
            var measure = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = AFM_1.AFM.isPreviousPeriodMeasureDefinition(measure);
            expect(result).toEqual(true);
        });
    });
    describe('isAttributeSortItem', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isAttributeSortItem(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isAttributeSortItem(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when attribute sort item is tested', function () {
            var sortItem = {
                attributeSortItem: {
                    direction: 'asc',
                    attributeIdentifier: 'a1'
                }
            };
            var result = AFM_1.AFM.isAttributeSortItem(sortItem);
            expect(result).toEqual(true);
        });
        it('should return false when measure sort item is tested', function () {
            var sortItem = {
                measureSortItem: {
                    direction: 'asc',
                    locators: [{
                            measureLocatorItem: {
                                measureIdentifier: 'm1'
                            }
                        }]
                }
            };
            var result = AFM_1.AFM.isAttributeSortItem(sortItem);
            expect(result).toEqual(false);
        });
    });
    describe('isMeasureSortItem', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isMeasureSortItem(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isMeasureSortItem(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when attribute sort item is tested', function () {
            var sortItem = {
                attributeSortItem: {
                    direction: 'asc',
                    attributeIdentifier: 'a1'
                }
            };
            var result = AFM_1.AFM.isMeasureSortItem(sortItem);
            expect(result).toEqual(false);
        });
        it('should return true when measure sort item is tested', function () {
            var sortItem = {
                measureSortItem: {
                    direction: 'asc',
                    locators: [{
                            measureLocatorItem: {
                                measureIdentifier: 'm1'
                            }
                        }]
                }
            };
            var result = AFM_1.AFM.isMeasureSortItem(sortItem);
            expect(result).toEqual(true);
        });
    });
    describe('isMeasureLocatorItem', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isMeasureLocatorItem(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isMeasureLocatorItem(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when attribute locator is tested', function () {
            var locatorItem = {
                attributeLocatorItem: {
                    attributeIdentifier: 'a1',
                    element: 'element'
                }
            };
            var result = AFM_1.AFM.isMeasureLocatorItem(locatorItem);
            expect(result).toEqual(false);
        });
        it('should return true when measure locator filter is tested', function () {
            var locatorItem = {
                measureLocatorItem: {
                    measureIdentifier: 'm1'
                }
            };
            var result = AFM_1.AFM.isMeasureLocatorItem(locatorItem);
            expect(result).toEqual(true);
        });
    });
    var expressionFilter = {
        value: 'MAQL'
    };
    var relativeDateFilter = {
        relativeDateFilter: {
            dataSet: {
                uri: '/gdc/mock/ds'
            },
            granularity: 'gram',
            from: -10,
            to: 0
        }
    };
    var absoluteDateFilter = {
        absoluteDateFilter: {
            dataSet: {
                uri: '/gdc/mock/ds'
            },
            from: '1',
            to: '2'
        }
    };
    var negativeAttributeFilter = {
        negativeAttributeFilter: {
            displayForm: {
                uri: '/gdc/mock/date'
            },
            notIn: ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
        }
    };
    var positiveAttributeFilter = {
        positiveAttributeFilter: {
            displayForm: {
                uri: '/gdc/mock/attribute'
            },
            "in": ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
        }
    };
    var measureValueFilter = {
        measureValueFilter: {
            measure: {
                uri: '/gdc/mock/date'
            }
        }
    };
    var rankingFilter = {
        rankingFilter: {
            measures: [{
                    uri: '/gdc/mock/date'
                }],
            operator: 'TOP',
            value: 3
        }
    };
    describe('isDateFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isDateFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isDateFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isDateFilter(relativeDateFilter);
            expect(result).toEqual(true);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isDateFilter(absoluteDateFilter);
            expect(result).toEqual(true);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isDateFilter(negativeAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isDateFilter(positiveAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isDateFilter(measureValueFilter);
            expect(result).toEqual(false);
        });
        it('should return false when ranking filter is tested', function () {
            var result = AFM_1.AFM.isDateFilter(rankingFilter);
            expect(result).toEqual(false);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isDateFilter(expressionFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isRelativeDateFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(relativeDateFilter);
            expect(result).toEqual(true);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(absoluteDateFilter);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(negativeAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(positiveAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(measureValueFilter);
            expect(result).toEqual(false);
        });
        it('should return false when ranking filter is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(rankingFilter);
            expect(result).toEqual(false);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isRelativeDateFilter(expressionFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isAbsoluteDateFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(relativeDateFilter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(absoluteDateFilter);
            expect(result).toEqual(true);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(negativeAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(positiveAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(measureValueFilter);
            expect(result).toEqual(false);
        });
        it('should return false when ranking filter is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(rankingFilter);
            expect(result).toEqual(false);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isAbsoluteDateFilter(expressionFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isAttributeFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(relativeDateFilter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(absoluteDateFilter);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(negativeAttributeFilter);
            expect(result).toEqual(true);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(positiveAttributeFilter);
            expect(result).toEqual(true);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(measureValueFilter);
            expect(result).toEqual(false);
        });
        it('should return false when ranking filter is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(rankingFilter);
            expect(result).toEqual(false);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isAttributeFilter(expressionFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isPositiveAttributeFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(relativeDateFilter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(absoluteDateFilter);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(negativeAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(positiveAttributeFilter);
            expect(result).toEqual(true);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(measureValueFilter);
            expect(result).toEqual(false);
        });
        it('should return false when ranking filter is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(rankingFilter);
            expect(result).toEqual(false);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isPositiveAttributeFilter(expressionFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isNegativeAttributeFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(relativeDateFilter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(absoluteDateFilter);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(negativeAttributeFilter);
            expect(result).toEqual(true);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(positiveAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(measureValueFilter);
            expect(result).toEqual(false);
        });
        it('should return false when ranking filter is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(rankingFilter);
            expect(result).toEqual(false);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isNegativeAttributeFilter(expressionFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isMeasureValueFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(relativeDateFilter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(absoluteDateFilter);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(negativeAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(positiveAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(measureValueFilter);
            expect(result).toEqual(true);
        });
        it('should return false when ranking filter is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(rankingFilter);
            expect(result).toEqual(false);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isMeasureValueFilter(expressionFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isRankingFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(relativeDateFilter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(absoluteDateFilter);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(negativeAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(positiveAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(measureValueFilter);
            expect(result).toEqual(false);
        });
        it('should return true when ranking filter is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(rankingFilter);
            expect(result).toEqual(true);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isRankingFilter(expressionFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isExpressionFilter', function () {
        it('should return false when null is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(relativeDateFilter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(absoluteDateFilter);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(negativeAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when positive attribute filter is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(positiveAttributeFilter);
            expect(result).toEqual(false);
        });
        it('should return false when measure value filter is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(measureValueFilter);
            expect(result).toEqual(false);
        });
        it('should return false when ranking filter is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(rankingFilter);
            expect(result).toEqual(false);
        });
        it('should return false when expression filter is tested', function () {
            var result = AFM_1.AFM.isExpressionFilter(expressionFilter);
            expect(result).toEqual(true);
        });
    });
});
