// @gooddata/typings/dist/src/tests/Execution.test.js
"use strict";
exports.__esModule = true;
// (C) 2007-2018 GoodData Corporation
var Execution_1 = require("../Execution");
var isAttributeHeaderItem = Execution_1.Execution.isAttributeHeaderItem;
var isMeasureHeaderItem = Execution_1.Execution.isMeasureHeaderItem;
var isTotalHeaderItem = Execution_1.Execution.isTotalHeaderItem;
var isMeasureGroupHeader = Execution_1.Execution.isMeasureGroupHeader;
var isAttributeHeader = Execution_1.Execution.isAttributeHeader;
describe('Execution type guards', function () {
    var attr = {
        attributeHeaderItem: {
            uri: '/uri',
            name: 'Name'
        }
    };
    var measure = {
        measureHeaderItem: {
            name: 'Name',
            order: 1
        }
    };
    var total = {
        totalHeaderItem: {
            name: 'Name',
            type: 'asdf'
        }
    };
    it('should recognize IResultHeaderItem subtypes', function () {
        expect(isAttributeHeaderItem(attr)).toEqual(true);
        expect(isMeasureHeaderItem(attr)).toEqual(false);
        expect(isTotalHeaderItem(attr)).toEqual(false);
        expect(isAttributeHeaderItem(measure)).toEqual(false);
        expect(isMeasureHeaderItem(measure)).toEqual(true);
        expect(isTotalHeaderItem(measure)).toEqual(false);
        expect(isAttributeHeaderItem(total)).toEqual(false);
        expect(isMeasureHeaderItem(total)).toEqual(false);
        expect(isTotalHeaderItem(total)).toEqual(true);
    });
    var measureGroup = {
        measureGroupHeader: {
            items: [
                {
                    measureHeaderItem: {
                        localIdentifier: 'm1',
                        name: 'Senseilevel',
                        format: '###'
                    }
                }
            ]
        }
    };
    var attributeHeader = {
        attributeHeader: {
            uri: '/uri',
            identifier: 'id',
            localIdentifier: 'a1',
            name: 'Year over year',
            formOf: {
                uri: '/uri2',
                identifier: 'id2',
                name: 'Date of birth'
            }
        }
    };
    it('should recognize IHeader subtypes', function () {
        expect(isMeasureGroupHeader(measureGroup)).toEqual(true);
        expect(isAttributeHeader(measureGroup)).toEqual(false);
        expect(isMeasureGroupHeader(attributeHeader)).toEqual(false);
        expect(isAttributeHeader(attributeHeader)).toEqual(true);
    });
});
