// @gooddata/typings/dist/src/tests/VisualizationInputs.test.js
"use strict";
exports.__esModule = true;
// (C) 2019-2020 GoodData Corporation
var VisualizationInput_1 = require("../VisualizationInput");
describe('VisualizationObject', function () {
    describe('isMeasure', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isMeasure(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isMeasure(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when visualization attribute is tested', function () {
            var attribute = {
                visualizationAttribute: {
                    localIdentifier: 'm1',
                    displayForm: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasure(attribute);
            expect(result).toEqual(false);
        });
        it('should return true when measure is tested', function () {
            var measure = {
                measure: {
                    localIdentifier: 'm1',
                    definition: {
                        measureDefinition: {
                            item: {
                                uri: '/gdc/mock/measure'
                            }
                        }
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasure(measure);
            expect(result).toEqual(true);
        });
    });
    describe('isAttribute', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isAttribute(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isAttribute(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when measure is tested', function () {
            var measure = {
                measure: {
                    localIdentifier: 'm1',
                    definition: {
                        measureDefinition: {
                            item: {
                                uri: '/gdc/mock/measure'
                            }
                        }
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isAttribute(measure);
            expect(result).toEqual(false);
        });
        it('should return true when visualization attribute is tested', function () {
            var attribute = {
                visualizationAttribute: {
                    localIdentifier: 'm1',
                    displayForm: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isAttribute(attribute);
            expect(result).toEqual(true);
        });
    });
    describe('isMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when simple measure definition is tested', function () {
            var measureDefinition = {
                measureDefinition: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(true);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measureDefinition = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when pop measure definition is tested', function () {
            var measureDefinition = {
                popMeasureDefinition: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measureDefinition = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
    });
    describe('isArithmeticMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isArithmeticMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isArithmeticMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measureDefinition = {
                measureDefinition: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isArithmeticMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return true when arithmetic measure definition is tested', function () {
            var measureDefinition = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isArithmeticMeasureDefinition(measureDefinition);
            expect(result).toEqual(true);
        });
        it('should return false when pop measure definition is tested', function () {
            var measureDefinition = {
                popMeasureDefinition: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isArithmeticMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measureDefinition = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isArithmeticMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
    });
    describe('isPopMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isPopMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isPopMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measureDefinition = {
                measureDefinition: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPopMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measureDefinition = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPopMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return true when pop measure definition is tested', function () {
            var measureDefinition = {
                popMeasureDefinition: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPopMeasureDefinition(measureDefinition);
            expect(result).toEqual(true);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measureDefinition = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
    });
    describe('isPreviousPeriodMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isPreviousPeriodMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isPreviousPeriodMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measureDefinition = {
                measureDefinition: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPreviousPeriodMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measureDefinition = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPreviousPeriodMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when pop measure definition is tested', function () {
            var measureDefinition = {
                popMeasureDefinition: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPreviousPeriodMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return true when previous period measure definition is tested', function () {
            var measureDefinition = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPreviousPeriodMeasureDefinition(measureDefinition);
            expect(result).toEqual(true);
        });
    });
    describe('isPositiveAttributeFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isPositiveAttributeFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isPositiveAttributeFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var filter = {
                negativeAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/date'
                    },
                    notIn: ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPositiveAttributeFilter(filter);
            expect(result).toEqual(false);
        });
        it('should return true when positive attribute filter is tested', function () {
            var filter = {
                positiveAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/attribute'
                    },
                    "in": ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isPositiveAttributeFilter(filter);
            expect(result).toEqual(true);
        });
    });
    describe('isAbsoluteDateFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isAbsoluteDateFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isAbsoluteDateFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when relative date filter is tested', function () {
            var filter = {
                relativeDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    granularity: 'GDC.time.year',
                    from: -1,
                    to: -1
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isAbsoluteDateFilter(filter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var filter = {
                absoluteDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    from: '2017-06-12',
                    to: '2018-07-11'
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isAbsoluteDateFilter(filter);
            expect(result).toEqual(true);
        });
    });
    describe('isRelativeDateFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isRelativeDateFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isRelativeDateFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when absolute date filter is tested', function () {
            var filter = {
                absoluteDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    from: 'beginning',
                    to: 'to end'
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isRelativeDateFilter(filter);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var filter = {
                relativeDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    granularity: 'GDC.time.year',
                    from: -1,
                    to: -1
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isRelativeDateFilter(filter);
            expect(result).toEqual(true);
        });
    });
    describe('isMeasureValueFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isMeasureValueFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isMeasureValueFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when measure value filter is tested', function () {
            var filter = {
                measureValueFilter: {
                    measure: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasureValueFilter(filter);
            expect(result).toEqual(true);
        });
        it('should return false when positive attribute filter is tested', function () {
            var filter = {
                positiveAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/attribute'
                    },
                    "in": ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isMeasureValueFilter(filter);
            expect(result).toEqual(false);
        });
    });
    describe('isRankingFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isRankingFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationInput_1.VisualizationInput.isRankingFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when measure value filter is tested', function () {
            var filter = {
                rankingFilter: {
                    measures: [{
                            uri: '/gdc/mock/date'
                        }],
                    operator: 'TOP',
                    value: 3
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isRankingFilter(filter);
            expect(result).toEqual(true);
        });
        it('should return false when positive attribute filter is tested', function () {
            var filter = {
                positiveAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/attribute'
                    },
                    "in": ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationInput_1.VisualizationInput.isRankingFilter(filter);
            expect(result).toEqual(false);
        });
    });
});
