// @gooddata/typings/dist/src/tests/VisualizationObject.test.js
"use strict";
exports.__esModule = true;
// (C) 2007-2020 GoodData Corporation
var VisualizationObject_1 = require("../VisualizationObject");
var isDateFilter = VisualizationObject_1.VisualizationObject.isDateFilter;
describe('VisualizationObject', function () {
    describe('isMeasure', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isMeasure(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isMeasure(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when visualization attribute is tested', function () {
            var attribute = {
                visualizationAttribute: {
                    localIdentifier: 'm1',
                    displayForm: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasure(attribute);
            expect(result).toEqual(false);
        });
        it('should return true when measure is tested', function () {
            var measure = {
                measure: {
                    localIdentifier: 'm1',
                    definition: {
                        measureDefinition: {
                            item: {
                                uri: '/gdc/mock/measure'
                            }
                        }
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasure(measure);
            expect(result).toEqual(true);
        });
    });
    describe('isVisualizationAttribute', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isVisualizationAttribute(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isVisualizationAttribute(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when measure is tested', function () {
            var measure = {
                measure: {
                    localIdentifier: 'm1',
                    definition: {
                        measureDefinition: {
                            item: {
                                uri: '/gdc/mock/measure'
                            }
                        }
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isVisualizationAttribute(measure);
            expect(result).toEqual(false);
        });
        it('should return true when visualization attribute is tested', function () {
            var attribute = {
                visualizationAttribute: {
                    localIdentifier: 'm1',
                    displayForm: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isVisualizationAttribute(attribute);
            expect(result).toEqual(true);
        });
    });
    describe('isMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when simple measure definition is tested', function () {
            var measureDefinition = {
                measureDefinition: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(true);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measureDefinition = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when pop measure definition is tested', function () {
            var measureDefinition = {
                popMeasureDefinition: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measureDefinition = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
    });
    describe('isArithmeticMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isArithmeticMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isArithmeticMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measureDefinition = {
                measureDefinition: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isArithmeticMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return true when arithmetic measure definition is tested', function () {
            var measureDefinition = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isArithmeticMeasureDefinition(measureDefinition);
            expect(result).toEqual(true);
        });
        it('should return false when pop measure definition is tested', function () {
            var measureDefinition = {
                popMeasureDefinition: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isArithmeticMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measureDefinition = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isArithmeticMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
    });
    describe('isPopMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isPopMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isPopMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measureDefinition = {
                measureDefinition: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPopMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measureDefinition = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPopMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return true when pop measure definition is tested', function () {
            var measureDefinition = {
                popMeasureDefinition: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPopMeasureDefinition(measureDefinition);
            expect(result).toEqual(true);
        });
        it('should return false when previous period measure definition is tested', function () {
            var measureDefinition = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
    });
    describe('isPreviousPeriodMeasureDefinition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isPreviousPeriodMeasureDefinition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isPreviousPeriodMeasureDefinition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when simple measure definition is tested', function () {
            var measureDefinition = {
                measureDefinition: {
                    item: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPreviousPeriodMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when arithmetic measure definition is tested', function () {
            var measureDefinition = {
                arithmeticMeasure: {
                    measureIdentifiers: ['/gdc/mock/measure'],
                    operator: 'sum'
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPreviousPeriodMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return false when pop measure definition is tested', function () {
            var measureDefinition = {
                popMeasureDefinition: {
                    measureIdentifier: 'm1',
                    popAttribute: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPreviousPeriodMeasureDefinition(measureDefinition);
            expect(result).toEqual(false);
        });
        it('should return true when previous period measure definition is tested', function () {
            var measureDefinition = {
                previousPeriodMeasure: {
                    measureIdentifier: 'm1',
                    dateDataSets: [{
                            dataSet: {
                                uri: '/gdc/mock/date'
                            },
                            periodsAgo: 1
                        }]
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPreviousPeriodMeasureDefinition(measureDefinition);
            expect(result).toEqual(true);
        });
    });
    describe('isAttributeFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isAttributeFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isAttributeFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when relative date filter is tested', function () {
            var filter = {
                relativeDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    granularity: 'GDC.time.year',
                    from: -1,
                    to: -1
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isAttributeFilter(filter);
            expect(result).toEqual(false);
        });
        it('should return true when negative attribute filter is tested', function () {
            var filter = {
                negativeAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/date'
                    },
                    notIn: ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isAttributeFilter(filter);
            expect(result).toEqual(true);
        });
        it('should return true when positive attribute filter is tested', function () {
            var filter = {
                positiveAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/attribute'
                    },
                    "in": ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isAttributeFilter(filter);
            expect(result).toEqual(true);
        });
    });
    describe('isDateFilter', function () {
        it('should return true for a VisualizationObjectAbsoluteDateFilter', function () {
            var absoluteDateFilter = {
                absoluteDateFilter: {
                    dataSet: null
                }
            };
            var result = isDateFilter(absoluteDateFilter);
            expect(result).toEqual(true);
        });
        it('should return true for a VisualizationObjectRelativeDateFilter', function () {
            var relativeDateFilter = {
                relativeDateFilter: {
                    dataSet: null,
                    granularity: ''
                }
            };
            var result = isDateFilter(relativeDateFilter);
            expect(result).toEqual(true);
        });
        it('should return false for a non VisualizationObjectDateFilter', function () {
            var attributeFilter = {
                positiveAttributeFilter: {
                    displayForm: null,
                    "in": []
                }
            };
            var result = isDateFilter(attributeFilter);
            expect(result).toEqual(false);
        });
    });
    describe('isPositiveAttributeFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isPositiveAttributeFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isPositiveAttributeFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when negative attribute filter is tested', function () {
            var filter = {
                negativeAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/date'
                    },
                    notIn: ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPositiveAttributeFilter(filter);
            expect(result).toEqual(false);
        });
        it('should return true when positive attribute filter is tested', function () {
            var filter = {
                positiveAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/attribute'
                    },
                    "in": ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isPositiveAttributeFilter(filter);
            expect(result).toEqual(true);
        });
    });
    describe('isAbsoluteDateFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isAbsoluteDateFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isAbsoluteDateFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when relative date filter is tested', function () {
            var filter = {
                relativeDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    granularity: 'GDC.time.year',
                    from: -1,
                    to: -1
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isAbsoluteDateFilter(filter);
            expect(result).toEqual(false);
        });
        it('should return true when absolute date filter is tested', function () {
            var filter = {
                absoluteDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    from: '2017-06-12',
                    to: '2018-07-11'
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isAbsoluteDateFilter(filter);
            expect(result).toEqual(true);
        });
    });
    describe('isRelativeDateFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isRelativeDateFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isRelativeDateFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when absolute date filter is tested', function () {
            var filter = {
                absoluteDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    from: 'beginning',
                    to: 'to end'
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isRelativeDateFilter(filter);
            expect(result).toEqual(false);
        });
        it('should return true when relative date filter is tested', function () {
            var filter = {
                relativeDateFilter: {
                    dataSet: {
                        uri: '/gdc/mock/date'
                    },
                    granularity: 'GDC.time.year',
                    from: -1,
                    to: -1
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isRelativeDateFilter(filter);
            expect(result).toEqual(true);
        });
    });
    describe('isAttribute', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isAttribute(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isAttribute(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when measure is tested', function () {
            var measure = {
                measure: {
                    localIdentifier: 'm1',
                    definition: {
                        measureDefinition: {
                            item: {
                                uri: '/gdc/mock/measure'
                            }
                        }
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isAttribute(measure);
            expect(result).toEqual(false);
        });
        it('should return true when visualization attribute is tested', function () {
            var attribute = {
                visualizationAttribute: {
                    localIdentifier: 'm1',
                    displayForm: {
                        uri: '/gdc/mock/measure'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isAttribute(attribute);
            expect(result).toEqual(true);
        });
    });
    describe('isMeasureValueFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isMeasureValueFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isMeasureValueFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when measure value filter is tested', function () {
            var filter = {
                measureValueFilter: {
                    measure: {
                        uri: '/gdc/mock/date'
                    }
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasureValueFilter(filter);
            expect(result).toEqual(true);
        });
        it('should return false when positive attribute filter is tested', function () {
            var filter = {
                positiveAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/attribute'
                    },
                    "in": ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isMeasureValueFilter(filter);
            expect(result).toEqual(false);
        });
    });
    describe('isRankingFilter', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isRankingFilter(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isRankingFilter(undefined);
            expect(result).toEqual(false);
        });
        it('should return true when measure value filter is tested', function () {
            var filter = {
                rankingFilter: {
                    measures: [{
                            uri: '/gdc/mock/date'
                        }],
                    operator: 'TOP',
                    value: 3
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isRankingFilter(filter);
            expect(result).toEqual(true);
        });
        it('should return false when positive attribute filter is tested', function () {
            var filter = {
                positiveAttributeFilter: {
                    displayForm: {
                        uri: '/gdc/mock/attribute'
                    },
                    "in": ['/gdc/mock/attribute/value_1', '/gdc/mock/attribute/value_2']
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isRankingFilter(filter);
            expect(result).toEqual(false);
        });
    });
    describe('isLocalIdentifierQualifier', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isLocalIdentifierQualifier(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isLocalIdentifierQualifier(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when object uri qualifier is tested', function () {
            var objUriQualifier = {
                uri: '/gdc/mock/measure'
            };
            var result = VisualizationObject_1.VisualizationObject.isLocalIdentifierQualifier(objUriQualifier);
            expect(result).toEqual(false);
        });
        it('should return true when object local qualifier is tested', function () {
            var localIdentifierQualifier = {
                localIdentifier: 'localId'
            };
            var result = VisualizationObject_1.VisualizationObject.isLocalIdentifierQualifier(localIdentifierQualifier);
            expect(result).toEqual(true);
        });
    });
    describe('isComparisonCondition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isComparisonCondition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isComparisonCondition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when a range condition is tested', function () {
            var rangeCondition = {
                range: {
                    operator: 'BETWEEN',
                    from: 10,
                    to: 100
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isComparisonCondition(rangeCondition);
            expect(result).toEqual(false);
        });
        it('should return true when a comparison condition is tested', function () {
            var comparisonCondition = {
                comparison: {
                    operator: 'EQUAL_TO',
                    value: 100
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isComparisonCondition(comparisonCondition);
            expect(result).toEqual(true);
        });
    });
    describe('isRangeCondition', function () {
        it('should return false when null is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isRangeCondition(null);
            expect(result).toEqual(false);
        });
        it('should return false when undefined is tested', function () {
            var result = VisualizationObject_1.VisualizationObject.isRangeCondition(undefined);
            expect(result).toEqual(false);
        });
        it('should return false when a comparison condition is tested', function () {
            var comparisonCondition = {
                comparison: {
                    operator: 'EQUAL_TO',
                    value: 100
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isRangeCondition(comparisonCondition);
            expect(result).toEqual(false);
        });
        it('should return true when a range condition is tested', function () {
            var rangeCondition = {
                range: {
                    operator: 'BETWEEN',
                    from: 10,
                    to: 100
                }
            };
            var result = VisualizationObject_1.VisualizationObject.isRangeCondition(rangeCondition);
            expect(result).toEqual(true);
        });
    });
});
